### DISCUSSION TITLE
Custom line graph rendering

### DISCUSSION THREAD

**Discussion by JanProchy** [katspaugh/wavesurfer.js#3461] [Q&A]
Hey,

first of all - thank you for this awesome library! 🏄‍♀️

Now the issue im trying to resolve. I need to render a line chart right bellow the wave. With support for cursor, regions and mini-map. So I'm reading all the discussions and issues here and few things cought my eye:

- Split Wave / Point Plot - https://www.jrweth.com/wavesurfer2/example/split-wave-point-plot/
  - seems promissing, but it's not supported any more in latest version?
- This guy was dealing with similiar issue - https://github.com/katspaugh/wavesurfer.js/issues/1255
  - but with the graph overlayed on top of wave
  - code was never released
- I checked the wavesurfer-multitrack - https://github.com/katspaugh/wavesurfer-multitrack
  - there is an issue that regions are not shared across canvases

The closest working example i found is the basic regions plugin example with `splitChannels: true`. Then if i would be able to somehow simulate the graph with envelope plugin or with own peaks and render function it would be done :-)

Here is a little preview of what I'm trying to achieve..
![image](https://github.com/katspaugh/wavesurfer.js/assets/23195511/0e4dea03-2177-46aa-bce4-fe7db8b4934d)

Do you guys have any tips/tricks/suggestions? :-)

Thanks! 🏄‍♀️

**Comment by katspaugh** [comment]
You could write a custom rendering function like here: https://wavesurfer.xyz/examples/?custom-render.js

To create a synced copy of wavesurfer:

```js
const wavesurfer = WaveSurfer.create({
  // ...
})

wavesurfer.on('decode', () => {
  const customRenderWavesurfer = WaveSurfer.create({
    container: wavesurfer.getWrapper(),
    media: wavesurfer.getMediaElement(),
    peaks:  wavesurfer.exportPeaks(),
    duration: wavesurfer.getDuration(),
    renderFunction: (channels, ctx) => {
      // ...
    },
  })
})
```

Here's a working demo: https://wavesurfer.xyz/examples/?85830fe8827040932066b40bdd03841f

**Reply by filipkoukal** [reply]
Hi, thanks for the suggestion, it works great but how would you go about overlaying multiple lines as shown in the preview from @JanProchy? I've tried using multiple channels to represent multiple charts with the `splitChannels` option, however this always puts the line charts underneath one another. I've found that there used to be an overlay option under `splitChannelsOptions`, which seems like exactly what I'm looking for, in older versions of Wavesurfer, but this functionality seems to have been removed in Wavesurfer 7. Would you have any tips or workarounds to achieve the old overlay functionality?

**Reply by katspaugh** [reply]
You can probably just offset the tracks below with a negative margin.
Like so: https://wavesurfer.xyz/examples/?5001303f60e060b71cbdfe56a7a43731

**Reply by filipkoukal** [reply]
That worked! Thank you so much!

**Reply by filipkoukal** [reply]
Hello once again. Sorry to resurrect this thread, but I've been wondering how you would go about adding a zoom plugin to this setup and have both Wavesurfer elements have synced zooming. I've tried registering a zoom plugin on the "main" Wavesurfer element and while that does work, the zooming itself isn't smooth for the lower Wavesurfer element (it seems like it's scrolling as well as zooming?). Once I finish zooming, the lower Wavesurfer then snaps to the correct zoom level and it all looks great, but my issue is with the behaviour while in the middle of zooming. If this can't be done with a workaround, I think it would be great to have built in to the library.
Here is my adjusted code from your example so you can see what I'm talking about: 

```
// Custom rendering function

import WaveSurfer from 'wavesurfer.js'
import ZoomPlugin from 'wavesurfer.js/dist/plugins/zoom';


const wavesurfer = WaveSurfer.create({
    container: document.body,
    waveColor: 'rgb(200, 0, 200)',
    progressColor: 'rgb(100, 0, 100)',
    url: '/examples/audio/demo.wav',
})

wavesurfer.registerPlugin(ZoomPlugin.create())

wavesurfer.on('ready', () => {
  const customRenderWavesurfer = WaveSurfer.create({
    container: wavesurfer.getWrapper(),
    media: wavesurfer.getMediaElement(),
    peaks:  wavesurfer.exportPeaks(),
    duration: wavesurfer.getDuration(),
    /**
     * Render a waveform as a squiggly line
     * @see https://css-tricks.com/making-an-audio-waveform-visualizer-with-vanilla-javascript/
     */
    renderFunction: (channels, ctx) => {
      const { width, height } = ctx.canvas
      const scale = channels[0].length / width
      const step = 10

      ctx.translate(0, height / 2)
      ctx.strokeStyle = ctx.fillStyle
      ctx.beginPath()

      for (let i = 0; i < width; i += step * 2) {
        const index = Math.floor(i * scale)
        const value = Math.abs(channels[0][index])
        let x = i
        let y = value * height

        ctx.moveTo(x, 0)
        ctx.lineTo(x, y)
        ctx.arc(x + step / 2, y, step / 2, Math.PI, 0, true)
        ctx.lineTo(x + step, 0)

        x = x + step
        y = -y
        ctx.moveTo(x, 0)
        ctx.lineTo(x, y)
        ctx.arc(x + step / 2, y, step / 2, Math.PI, 0, false)
        ctx.lineTo(x + step, 0)
      }

      ctx.stroke()
      ctx.closePath()
    },
  })
})

wavesurfer.on('interaction', () => {
  wavesurfer.play()
})

```

**Reply by katspaugh** [reply]
You could probably add `customRenderWavesurfer.registerPlugin(ZoomPlugin.create())` so that it zooms on each scroll.

**Reply by filipkoukal** [reply]
I tried that, but sadly it doesn't work the way I would want. Zooming on the top main Wavesurfer element still causes the same behaviour on the lower one and it's now possible to cause them to desync when zooming on the main one first and the lower one after. This approach works pretty well when the Wavesurfer elements are stacked on top of one another with negative margin though, which is pretty nice.

**Reply by katspaugh** [reply]
Then perhaps something like this:

```js
  wavesurfer.on('zoom', (minPxPerSec) => {
    customRenderWavesurfer.setOptions({ minPxPerSec })
  })
```

**Reply by filipkoukal** [reply]
Yes! That seems to be working exactly how I would want! Thank you so much!
